### PR TITLE
Iter13

### PR DESCRIPTION
**Pull request by justEngineer**

### CODE REVIEW

**Review by arttet — CHANGES_REQUESTED** [review]
Please fix the `CodeCov` action

File: internal/database/database.go [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+	//connect, err := pgx.Connect(context.Background(), os.Getenv("DATABASE_URL"))
+	_, err = connect.Exec(ctx, createTablesSQL)
+	return &db, err
+}
+
+func (d *Database) Ping() error {
+	// ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+	// defer cancel()
+	// err := d.Connections.Ping(ctx)
+	err := d.Connections.Ping(context.Background())
```
[the remark this answers is not in the document]
Comment (justEngineer) [reply internal/database/database.go]: done

File: internal/database/database.go [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+)
+
+type Database struct {
+	Connections *pgxpool.Pool
+}
+
+const (
+	createTablesSQL = `
+	CREATE TABLE IF NOT EXISTS gauge_metrics (
+		id      TEXT PRIMARY KEY,
```
[the remark this answers is not in the document]
Comment (justEngineer) [reply internal/database/database.go]: replaced with VARCHAR (255)

File: internal/database/database.go [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+	config "github.com/justEngineer/go-metrics-service/internal/http/server/config"
+	"github.com/justEngineer/go-metrics-service/internal/storage"
+)
+
+type Database struct {
+	Connections *pgxpool.Pool
+}
+
+const (
+	createTablesSQL = `
```
[the remark this answers is not in the document]
Comment (justEngineer) [reply internal/database/database.go]: done

File: internal/database/database.go [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+	VALUES ($1, $2)
+	ON CONFLICT (id) DO UPDATE
+	SET value = counter_metrics.value + EXCLUDED.value`
+
+	selectGaugeSQL = `SELECT value FROM gauge_metrics WHERE id = $1`
+
+	selectCounterSQL = `SELECT value FROM counter_metrics WHERE id = $1`
+)
+
+var retryTimeouts = [4]int{0, 1, 2, 3}
```
[the remark this answers is not in the document]
Comment (justEngineer) [reply internal/database/database.go]: thank you for the advice. done

File: internal/database/database.go [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+	for i := 0; i < len(retryTimeouts); i++ {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			err = f(ctx, key, value)
+			if err == nil {
+				return err
+			} else {
+				time.Sleep(time.Duration(retryTimeouts[i]) * time.Second)
```
[the remark this answers is not in the document]
Comment (justEngineer) [reply internal/database/database.go]: replaced with separate function "executeWithBackoff"

**Review by arttet — APPROVED** [review]

**Pull request merged**